Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,345 @@
+/-
+Copyright (c) 2024 Miguel Marco. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Miguel Marco
+-/
+import Mathlib.Data.Set.Function
+import Mathlib.Data.Set.Functor
+import Mathlib.Lean.Expr.ExtraRecognizers
+
+/-!
+# Subsets as subtypes
+
+This file defines the function `setRestrict` that converts sets in a type, to sets in a subtype
+(as the elements in the subtype whose lift lives in the set), and some notation and simplification
+lemmas for convenience.
+
+More precisely, given `α : Type` and `A : Set α`, `setRestrict A : Set α → Set A` takes
+some `B : Set α` and produces some `Set ↑A` whose elements are `{ x : ↑A | ↑x ∈ B}`.
+
+It also defines the notation `A ↓∩ B` for `setRestrict A B`.
+
+## Implementation
+
+The simplification lemmas are designed to simplify the expression appearing inside the `↓∩`
+operation, if possible.
+
+## Tags
+
+subsets
+-/
+
+open Lean PrettyPrinter Delaborator SubExpr in
+/--
+Sets of a subtype coerced to the ambient type are denoted with `↑`.
+-/
+@[delab app.Set.image]
+def delab_set_image_subtype : Delab := do
+  let #[α, _, f, _] := (← getExpr).getAppArgs | failure
+  guard <| f.isAppOfArity ``Subtype.val 2
+  let some _ := α.coeTypeSet? | failure
+  let e ← withAppArg delab
+  `(↑$e)
+
+open Set
+
+universe u
+
+variable {α β : Type u} {A B C: Set α} {D E : Set ↑A}
+variable {S : Set (Set α)} {T : Set (Set ↑A)} {i : β → Set α} {j : β → Set ↑A}
+
+namespace Subset
+
+/--
+Given two sets `A` and `B`, `setRestrict A B` is the set of `↑A` formed by the elements
+whose value is in `B`.
+-/
+def setRestrict (A B : Set α) : Set ↑A := (↑) ⁻¹' B
+
+/--
+`A ↓∩ B` denotes `restrict A B`.
+-/
+infixl:67 " ↓∩ "  => setRestrict
+
+@[simp]
+lemma mem_setRestrict_iff (x : A) : x ∈  (A ↓∩ B) ↔ ↑x ∈ B := by rfl
+
+@[simp]
+lemma setRestrict_empty : A ↓∩ (∅ : Set α) = (∅ : Set A) := rfl
+
+@[simp]
+lemma setRestrict_self : A ↓∩ A = univ := by
+  ext x
+  simp only [mem_setRestrict_iff, Subtype.coe_prop, mem_univ]
+
+@[simp]
+lemma setRestrict_univ : A ↓∩ (univ : Set α) = (univ : Set A) := rfl
+
+@[simp]
+lemma setRestrict_union : A ↓∩ (B ∪ C) = (A ↓∩ B) ∪ (A ↓∩ C) := rfl
+
+@[simp]
+lemma setRestrict_inter : A ↓∩ (B ∩ C) = (A ↓∩ B) ∩ (A ↓∩ C) := rfl
+
+@[simp]
+lemma set_restrit_compl : A ↓∩ Bᶜ = (A ↓∩ B)ᶜ := by
+  apply Eq.refl
+
+lemma setRestrict_eq_univ_of_subset (h : A ⊆ B) : A ↓∩ B = univ := by
+  ext x
+  simp only [mem_setRestrict_iff, mem_univ, iff_true]
+  exact h x.2
+
+lemma restrict_subsetRestrict_iff : A ∩ B ⊆ A ∩ C ↔ A ↓∩ B ⊆ A ↓∩ C := by
+  constructor
+  · rintro h ⟨x,hxA⟩ hx
+    specialize h ?_
+    · exact x
+    · exact ⟨hxA,hx⟩
+    exact h.2
+  · rintro h x ⟨hxA,hxB⟩
+    constructor
+    · exact hxA
+    · specialize h ?_
+      · exact ⟨x,hxA⟩
+      · exact hxB
+      exact h
+
+@[simp]
+lemma setRestrict_eq_iff : A ↓∩ B = A ↓∩ C ↔ A ∩ B = A ∩ C := by
```

Comment:
Like `restrict_subsetRestrict_iff`, this shouldn't be a `simp` lemma. I don't mind the direction of the `iff` too much, but it should at least be consistent with `restrict_subsetRestrict_iff`.
```suggestion
lemma setRestrict_eq_iff : A ↓∩ B = A ↓∩ C ↔ A ∩ B = A ∩ C := by
```